File: backend/trading/routers/executions.py
"""Executions and transactions: CRUD, Flex fetch, IB fetch, performance."""

import asyncio
import logging
from typing import Any, Dict, List, Optional

from fastapi import APIRouter, Body, HTTPException, Query, Request

from src.portfolio.services.executions_fetch_flex import (
    fetch_flex_trades_and_upsert_executions,
    upsert_executions_from_uploaded_flex_xml,
)
from src.portfolio.services.transactions_fetch import fetch_cash_transactions_from_flex
from src.monitor.reader import (
    write_account_executions_to_db,
    insert_one_execution,
    update_one_execution,
    delete_one_execution,
)

logger = logging.getLogger(__name__)

router = APIRouter(tags=["executions"])


def _fmt_db_id_list(ids: Any, *, max_show: int = 48) -> str:
    """Format ``executions_raw_tws_id`` lists for API message text (truncate long lists)."""
    if not ids:
        return "none"
    try:
        nums = [int(x) for x in list(ids)]
    except (TypeError, ValueError):
        return "none"
    if len(nums) <= max_show:
        return "[" + ", ".join(str(i) for i in nums) + "]"
    head = nums[:max_show]
    return "[" + ", ".join(str(i) for i in head) + f", … +{len(nums) - max_show} more]"


def _publish_tws_fetch_system_message(
    config: dict,
    *,
    ok: bool,
    title: str,
    message: str,
    reason: Optional[str] = None,
    level: Optional[str] = None,
) -> None:
    """Best-effort Redis message center (Monitor materializes for SSE)."""
    try:
        import redis as redis_mod

        from src.bifrost.message_center import (
            build_portfolio_tws_executions_fetch_event,
            publish_system_message_event,
        )
        from src.core.redis_url import effective_redis_dict, format_redis_url

        url = format_redis_url(effective_redis_dict(config, default_db=0))
        if not url:
            return
        r = redis_mod.from_url(url, decode_responses=True)
        try:
            ev = build_portfolio_tws_executions_fetch_event(
                ok=ok, title=title, message=message, reason=reason, level=level
            )
            publish_system_message_event(r, ev)
        finally:
            r.close()
    except Exception as e:
        logger.debug("tws fetch message center publish failed: %s", e)


@router.get("/executions")
def get_executions(
    request: Request,
    since_ts: Optional[float] = Query(None, description="Filter executions with time >= this (Unix s)"),
    until_ts: Optional[float] = Query(None, description="Filter executions with time <= this"),
    account_id: Optional[str] = Query(None, description="Filter by account ID"),
    limit: int = Query(200, ge=0, le=10000, description="Max rows to return; 0 = no limit"),
    include_opt_pairs: bool = Query(False, description="Include C<>P pairing"),
    strategy_opportunity_id: Optional[int] = Query(None, description="Filter by strategy opportunity ID"),
    strategy_instance_id: Optional[int] = Query(None, description="Filter by strategy instance ID"),
    source_scope: Optional[str] = Query(
        None,
        description="Optional: all (default, full account_executions) | performance_book (account_executions_final) | on_the_fly (account_executions_fly: TWS not covered by final, no BAG) | tws_raw (executions_raw_tws only, synthetic negative account_executions_id)",
    ),
) -> Dict[str, Any]:
    """Account-level executions/trades (R-A2). If include_opt_pairs=true: returns paired_execution_ids and opt_pairs."""
    reader = request.app.state.reader
    effective_limit: Optional[int] = limit if limit > 0 else None
    if include_opt_pairs:
        return reader.get_executions_with_opt_pairs(
            since_ts=since_ts,
            until_ts=until_ts,
            account_id=account_id,
            limit=effective_limit or 5000,
            strategy_opportunity_id=strategy_opportunity_id,
            strategy_instance_id=strategy_instance_id,
            source_scope=source_scope,
        )
    items = reader.get_executions(
        since_ts=since_ts,
        until_ts=until_ts,
        account_id=account_id,
        limit=effective_limit,
        strategy_opportunity_id=strategy_opportunity_id,
        strategy_instance_id=strategy_instance_id,
        source_scope=source_scope,
    )
    return {"executions": items}


@router.get("/executions/position-attribution")
def get_position_attribution(
    request: Request,
    account_id: Optional[str] = Query(None, description="Filter by account ID"),
    sec_type: Optional[str] = Query(None, description="Filter by sec_type (e.g. OPT, STK)"),
) -> Dict[str, Any]:
    """Position x Instance attribution (net-estimated). Returns one row per (position, instance)."""
    reader = request.app.state.reader
    items = reader.get_position_instance_attribution(
        account_id=account_id,
        sec_type_filter=sec_type,
    )
    return {"attributions": items}


@router.get("/executions/link-candidates")
def get_executions_link_candidates(
    request: Request,
    account_id: str = Query(..., description="IB account id"),
    contract_key: Optional[str] = Query(None, description="Exact contract_key match (preferred)"),
    symbol: Optional[str] = Query(None),
    expiry: Optional[str] = Query(None, description="Option expiry (any format; used if contract_key yields no rows)"),
    strike: Optional[float] = Query(None),
    option_right: Optional[str] = Query(None, description="C or P"),
    limit: int = Query(200, ge=1, le=500),
) -> Dict[str, Any]:
    """Existing account_executions rows to link strategy attribution (no insert)."""
    reader = request.app.state.reader
    if not (contract_key and contract_key.strip()) and (
        not (symbol and symbol.strip()) or strike is None or expiry is None or str(expiry).strip() == ""
    ):
        return {
            "executions": [],
            "error": "Provide contract_key, or symbol+expiry+strike for fallback matching.",
        }
    items = reader.get_executions_for_strategy_link(
        account_id=account_id.strip(),
        contract_key=(contract_key or "").strip() or None,
        symbol=(symbol or "").strip() or None,
        expiry=expiry,
        strike=strike,
        option_right=(option_right or "").strip() or None,
        limit=limit,
    )
    return {"executions": items}


@router.get("/executions/freshness")
def get_executions_freshness(request: Request) -> Dict[str, Any]:
    """Execution data freshness per (account_id, source). Latest exec_time and days_since_latest."""
    reader = request.app.state.reader
    items = reader.get_executions_freshness()
    return {"items": items}


@router.get("/performance")
def get_performance(
    request: Request,
    since_ts: Optional[float] = Query(None),
    until_ts: Optional[float] = Query(None),
    account_id: Optional[str] = Query(None),
    granularity: str = Query("day", description="day | week | month"),
    strategy_opportunity_id: Optional[int] = Query(None, description="Filter by strategy opportunity ID"),
    strategy_instance_id: Optional[int] = Query(None, description="Filter by strategy instance ID"),
    source_scope: str = Query(
        "performance_book",
        description="performance_book (default, account_executions_final) | on_the_fly (account_executions_fly)",
    ),
    summary_only: bool = Query(
        False,
        description="With strategy_instance_id only: return summary via one SQL aggregate (fast)",
    ),
) -> Dict[str, Any]:
    """Performance stats and calendar PnL. Default source_scope=performance_book reads account_executions_final (flex+journal only)."""
    reader = request.app.state.reader
    if summary_only:
        if strategy_instance_id is None:
            raise HTTPException(status_code=400, detail="summary_only requires strategy_instance_id")
        if (account_id is not None and str(account_id).strip()) or strategy_opportunity_id is not None:
            raise HTTPException(
                status_code=400,
                detail="summary_only allows only strategy_instance_id (no account_id / opportunity filter)",
            )
        out = reader.get_performance_instance_summary(
            strategy_instance_id=strategy_instance_id,
            since_ts=since_ts,
            until_ts=until_ts,
        )
        return out
    out = reader.get_performance_stats(
        since_ts=since_ts,
        until_ts=until_ts,
        account_id=account_id,
        granularity=granularity,
        strategy_opportunity_id=strategy_opportunity_id,
        strategy_instance_id=strategy_instance_id,
        source_scope=source_scope,
    )
    return out


@router.get("/transactions")
def get_transactions(
    request: Request,
    since_ts: Optional[float] = Query(None),
    until_ts: Optional[float] = Query(None),
    account_id: Optional[str] = Query(None),
    limit: int = Query(500),
) -> Dict[str, Any]:
    """List account_transactions (Flex cash transactions) for Transfer & Pay page."""
    reader = request.app.state.reader
    items = reader.get_transactions(since_ts=since_ts, until_ts=until_ts, account_id=account_id, limit=limit)
    return {"transactions": items}


@router.post("/transactions/fetch")
def post_transactions_fetch(request: Request, body: Dict[str, Any] = Body(default=None)) -> Dict[str, Any]:
    """Fetch cash transactions from IB Flex and upsert into account_transactions."""
    reader = request.app.state.reader
    control_via_db = request.app.state.control_via_db
    return fetch_cash_transactions_from_flex(reader, control_via_db, body)


@router.post("/executions/fetch-flex")
def post_executions_fetch_flex(request: Request, body: Dict[str, Any] = Body(default=None)) -> Dict[str, Any]:
    """Fetch executions/trades from IB Flex (Trades report) and upsert into account_executions."""
    reader = request.app.state.reader
    control_via_db = request.app.state.control_via_db
    return fetch_flex_trades_and_upsert_executions(reader, control_via_db, body)


@router.post("/executions/fetch-flex-upload")
def post_executions_fetch_flex_upload(request: Request, body: Dict[str, Any] = Body(...)) -> Dict[str, Any]:
    """Upload Flex Trades XML and upsert into account_executions. Body: { "xml": "<FlexStatement ...>...</FlexStatement>" }"""
    control_via_db = request.app.state.control_via_db
    reader = request.app.state.reader
    raw_xml = (body.get("xml") or "").strip()
    cfg = getattr(reader, "_config", None)
    return upsert_executions_from_uploaded_flex_xml(control_via_db, raw_xml, config=cfg)


@router.post("/executions")
def post_execution(request: Request, body: Dict[str, Any] = Body(...)) -> Dict[str, Any]:
    """Add one execution record manually (history). body: account_id, time, symbol, sec_type, side, quantity, price; optional fields."""
    control_via_db = request.app.state.control_via_db
    if not control_via_db:
        return {"ok": False, "error": "PostgreSQL is required to write account_executions.", "account_executions_id": None}
    new_account_executions_id = insert_one_execution(control_via_db, body)
    if new_account_executions_id is None:
        return {"ok": False, "error": "Failed to add execution (required fields: symbol, quantity, price).", "account_executions_id": None}
    return {"ok": True, "account_executions_id": new_account_executions_id, "message": "Execution record added."}


def _parse_account_executions_path_id(execution_id: str) -> int:
    try:
        return int(str(execution_id).strip())
    except (TypeError, ValueError) as e:
        raise HTTPException(status_code=422, detail="Invalid execution id") from e


@router.put("/executions/{execution_id}")
def put_execution(request: Request, execution_id: str, body: Dict[str, Any] = Body(...)) -> Dict[str, Any]:
    """Update one execution by account_executions_id (manual correction). Negative ids = TWS raw rows."""
    eid = _parse_account_executions_path_id(execution_id)
    control_via_db = request.app.state.control_via_db
    if not control_via_db:
        return {"ok": False, "error": "PostgreSQL is required to write account_executions."}
    if update_one_execution(control_via_db, eid, body):
        return {"ok": True, "message": "Execution record updated."}
    return {"ok": False, "error": "Update failed (account_executions_id missing or database error)."}


@router.delete("/executions/{execution_id}")
def delete_execution(request: Request, execution_id: str) -> Dict[str, Any]:
    """Delete one execution by account_executions_id. Negative ids = TWS raw rows."""
    eid = _parse_account_executions_path_id(execution_id)
    control_via_db = request.app.state.control_via_db
    if not control_via_db:
        return {"ok": False, "error": "PostgreSQL is required to write account_executions."}
    if delete_one_execution(control_via_db, eid):
        return {"ok": True, "message": "Execution record deleted."}
    return {"ok": False, "error": "Delete failed (account_executions_id missing or database error)."}


@router.post("/executions/fetch")
async def post_executions_fetch(
    request: Request,
    days: int = Query(1, ge=1, le=7, description="1=today, 3=last 3 days, 7=last 7 days (TWS Trade Log)"),
) -> Dict[str, Any]:
    """Fetch executions from IB via IB Gateway and write to account_executions."""
    app = request.app
    reader = app.state.reader
    cfg = reader._config
    control_via_db = app.state.control_via_db
    if not control_via_db:
        return {"ok": False, "error": "PostgreSQL is required to write account_executions.", "count": 0}
    if not getattr(app.state, "monitor_enabled", True):
        return {"ok": False, "error": "Monitor stopped; cannot fetch executions.", "count": 0}
    gw = getattr(app.state, "ib_operator_client", None)
    if gw is None:
        return {"ok": False, "error": "IB Gateway client is not configured.", "count": 0}
    env = await gw.request_async(
        "fetch_executions",
        {"days": days, "account_slot": "primary"},
        caller="trading_executions_fetch",
    )
    if not env.get("ok"):
        err = str(env.get("error") or "IB gateway error")
        await asyncio.to_thread(
            _publish_tws_fetch_system_message,
            cfg,
            ok=False,
            title="TWS executions fetch failed",
            message=f"Primary slot: {err}",
            reason=err,
            level="error",
        )
        return {"ok": False, "error": err, "count": 0, "days": days, "fetched_primary": 0, "fetched_secondary": 0, "fetched_total": 0}
    data = env.get("data") or {}
    primary_execs = list(data.get("executions") or [])
    fetched_primary = len(primary_execs)
    all_execs = primary_execs
    fetched_secondary = 0
    secondary_error: Optional[str] = None
    from src.app.config import get_effective_ib_config

    try:
        ibc = get_effective_ib_config(reader._config)
        if (ibc.get("ib2_host") or "").strip():
            env2 = await gw.request_async(
                "fetch_executions",
                {"days": days, "account_slot": "secondary"},
                caller="trading_executions_fetch",
            )
            if env2.get("ok"):
                d2 = env2.get("data") or {}
                ex2 = list(d2.get("executions") or [])
                fetched_secondary = len(ex2)
                if ex2:
                    all_execs = (all_execs or []) + ex2
            else:
                secondary_error = str(env2.get("error") or "secondary fetch failed")
                logger.warning("executions/fetch secondary: %s", secondary_error)
    except Exception as e2:
        logger.warning("executions/fetch secondary check: %s", e2)
        secondary_error = str(e2)
    fetched_total = len(all_execs)
    if not all_execs:
        msg = (
            f"IB returned no executions (range: last {days} day(s); for a multi-day range ensure TWS Trade Log includes those days)."
        )
        if secondary_error:
            msg = f"{msg} Secondary slot error: {secondary_error}"
        detail_parts = [
            f"days={days}",
            f"fetched_primary={fetched_primary}",
            f"fetched_secondary={fetched_secondary}",
            f"fetched_total=0",
        ]
        if secondary_error:
            detail_parts.append(f"secondary_error={secondary_error}")
        await asyncio.to_thread(
            _publish_tws_fetch_system_message,
            cfg,
            ok=True,
            title="TWS executions fetch: no rows",
            message=msg + " " + " ".join(detail_parts),
            reason=None,
            level="warning",
        )
        out: Dict[str, Any] = {
            "ok": True,
            "message": msg,
            "count": 0,
            "days": days,
            "fetched_primary": fetched_primary,
            "fetched_secondary": fetched_secondary,
            "fetched_total": 0,
        }
        if secondary_error:
            out["secondary_error"] = secondary_error
        return out

    stats_out: Dict[str, Any] = {}
    if not write_account_executions_to_db(control_via_db, all_execs, stats_out=stats_out):
        await asyncio.to_thread(
            _publish_tws_fetch_system_message,
            cfg,
            ok=False,
            title="TWS executions write failed",
            message="PostgreSQL write failed after IB returned executions.",
            reason="write_account_executions_to_db",
            level="error",
        )
        return {
            "ok": False,
            "error": "Failed to write account_executions.",
            "count": 0,
            "days": days,
            "fetched_primary": fetched_primary,
            "fetched_secondary": fetched_secondary,
            "fetched_total": fetched_total,
        }

    ins = int(stats_out.get("tws_raw_inserted") or 0)
    skip = int(stats_out.get("tws_raw_skipped_duplicate") or 0)
    missing_raw = bool(stats_out.get("tws_raw_missing_table"))
    ins_ids = list(stats_out.get("tws_raw_inserted_ids") or [])
    upd_ids = list(stats_out.get("tws_raw_updated_ids") or [])
    sk_ids = list(stats_out.get("tws_raw_skipped_ids") or [])
    msg = (
        f"Fetched {fetched_total} execution(s) from IB (primary {fetched_primary}, secondary {fetched_secondary}). "
        f"executions_raw_tws: inserted {ins} row(s), ids {_fmt_db_id_list(ins_ids)}; "
        f"updated {len(upd_ids)} row(s), ids {_fmt_db_id_list(upd_ids)} (TWS path is insert-or-skip, usually 0); "
        f"skipped duplicate exec_id: {skip}, existing row ids {_fmt_db_id_list(sk_ids)}."
    )
    if missing_raw:
        msg += " (executions_raw_tws missing or unavailable; stats may be incomplete.)"
    if secondary_error:
        msg += f" Secondary slot error (merged primary only): {secondary_error}"

    await asyncio.to_thread(
        _publish_tws_fetch_system_message,
        cfg,
        ok=True,
        title="TWS executions imported",
        message=msg,
        reason=None,
        level="success",
    )

    result: Dict[str, Any] = {
        "ok": True,
        "count": fetched_total,
        "days": days,
        "fetched_primary": fetched_primary,
        "fetched_secondary": fetched_secondary,
        "fetched_total": fetched_total,
        "tws_raw_inserted": ins,
        "tws_raw_skipped_duplicate": skip,
        "tws_raw_missing_table": missing_raw,
        "tws_raw_inserted_ids": ins_ids,
        "tws_raw_updated_ids": upd_ids,
        "tws_raw_skipped_ids": sk_ids,
        "message": msg,
    }
    if secondary_error:
        result["secondary_error"] = secondary_error
    return result

File: frontend/src/api/trading/executions.ts
import type {
  ControlResponse,
  Execution,
  ExecutionsFetchTwsResponse,
  ExecutionsResponse,
  ExecutionsResponseWithPairs,
  ExecutionsFreshnessResponse,
  ExecutionsFlexUploadResponse,
  OptionStockLinkRow,
  OptionStockLinkSummary,
  PositionAttributionResponse,
} from '../../types'
import { getTradingApiBase, joinServiceBase } from '../shared/apiRouting'

function apiBase(): string {
  return getTradingApiBase()
}

function tradingUrl(path: string): string {
  return joinServiceBase(apiBase(), path)
}

/** R-A2: API fetches executions from IB via IB Operator and writes to DB. days: 1=today, 3=3d, 7=7d */
export async function postExecutionsFetch(days: 1 | 3 | 7 = 1): Promise<ExecutionsFetchTwsResponse> {
  const params = new URLSearchParams({ days: String(days) })
  const r = await fetch(tradingUrl(`/executions/fetch?${params}`), { method: 'POST' })
  const j = await r.json().catch(() => ({}))
  return {
    ...j,
    ok: r.ok && j.ok !== false,
    error: j.error || (r.ok ? undefined : r.statusText),
    count: j.count,
  }
}

/** R-A2: Fetch executions via IB Flex Trades report and write to DB. Range from backend (Flex + Settings). */
export async function postExecutionsFetchFlex(
  body?: { from_date?: string; to_date?: string }
): Promise<
  ControlResponse & {
    count?: number
    raw_count?: number
    updated_accounts?: number
    range_mode?: string
    range_days?: number | null
    range_from?: string | null
    range_to?: string | null
    last_flex_date_after?: string | null
    data_from?: string | null
    data_to?: string | null
    by_account?: number
    by_account_counts?: number[]
    per_query?: Array<{
      role?: string
      query_id: string
      label?: string | null
      rows?: number
      data_from?: string | null
      data_to?: string | null
    }>
  }
> {
  const r = await fetch(tradingUrl('/executions/fetch-flex'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body ?? {}),
  })
  const j = await r.json().catch(() => ({}))
  return {
    ...j,
    ok: r.ok && j.ok !== false,
    error: j.error || (r.ok ? undefined : r.statusText),
    count: j.count,
    updated_accounts: j.updated_accounts,
    raw_count: j.raw_count,
    range_mode: j.range_mode,
    range_days: j.range_days,
    range_from: j.range_from,
    range_to: j.range_to,
    last_flex_date_after: j.last_flex_date_after,
    data_from: j.data_from,
    data_to: j.data_to,
    by_account: j.by_account,
    by_account_counts: j.by_account_counts,
    per_query: j.per_query,
  }
}

/** GET /executions/position-attribution: Position×Instance net-estimated attribution. */
export async function fetchPositionAttribution(
  account_id?: string,
  sec_type?: string,
): Promise<PositionAttributionResponse> {
  const params = new URLSearchParams()
  if (account_id) params.set('account_id', account_id)
  if (sec_type) params.set('sec_type', sec_type)
  const r = await fetch(tradingUrl(`/executions/position-attribution?${params}`))
  if (!r.ok) throw new Error(r.statusText)
  return r.json()
}

/** GET /executions/link-candidates: existing rows to attach strategy (no insert). */
export async function fetchExecutionLinkCandidates(params: {
  account_id: string
  contract_key?: string
  symbol?: string
  expiry?: string
  strike?: number
  option_right?: string
  limit?: number
}): Promise<{ executions: Execution[]; error?: string }> {
  const q = new URLSearchParams()
  q.set('account_id', params.account_id.trim())
  if (params.contract_key?.trim()) q.set('contract_key', params.contract_key.trim())
  if (params.symbol?.trim()) q.set('symbol', params.symbol.trim())
  if (params.expiry != null && String(params.expiry).trim() !== '') q.set('expiry', String(params.expiry).trim())
  if (params.strike != null && Number.isFinite(params.strike)) q.set('strike', String(params.strike))
  if (params.option_right?.trim()) q.set('option_right', params.option_right.trim().slice(0, 1))
  if (params.limit != null) q.set('limit', String(params.limit))
  const r = await fetch(tradingUrl(`/executions/link-candidates?${q}`))
  const j = await r.json().catch(() => ({}))
  if (!r.ok) {
    return { executions: [], error: (j as { error?: string }).error || r.statusText }
  }
  return { executions: (j as { executions?: Execution[] }).executions ?? [], error: (j as { error?: string }).error }
}

/** source_scope: performance_book → account_executions_final; on_the_fly → account_executions_fly; tws_raw → executions_raw_tws. */
export async function fetchExecutions(
  since_ts?: number,
  until_ts?: number,
  limit = 200,
  include_opt_pairs = false,
  strategy_opportunity_id?: number,
  strategy_instance_id?: number,
  source_scope?: 'performance_book' | 'on_the_fly' | 'tws_raw',
): Promise<ExecutionsResponse | ExecutionsResponseWithPairs> {
  const params = new URLSearchParams()
  if (since_ts != null) params.set('since_ts', String(since_ts))
  if (until_ts != null) params.set('until_ts', String(until_ts))
  params.set('limit', String(limit))
  if (include_opt_pairs) params.set('include_opt_pairs', 'true')
  if (strategy_opportunity_id != null) params.set('strategy_opportunity_id', String(strategy_opportunity_id))
  if (strategy_instance_id != null) params.set('strategy_instance_id', String(strategy_instance_id))
  if (source_scope) params.set('source_scope', source_scope)
  const r = await fetch(tradingUrl(`/executions?${params}`))
  if (!r.ok) throw new Error(r.statusText)
  return r.json()
}

/** GET /executions/freshness: latest exec_time per (account_id, source) from account_executions. */
export async function fetchExecutionsFreshness(): Promise<ExecutionsFreshnessResponse> {
  const r = await fetch(tradingUrl('/executions/freshness'))
  if (!r.ok) throw new Error(r.statusText)
  return r.json()
}

/** POST /executions/fetch-flex-upload: upsert executions from uploaded Flex Trades XML. */
export async function postExecutionsFetchFlexUpload(xml: string): Promise<ExecutionsFlexUploadResponse> {
  const r = await fetch(tradingUrl('/executions/fetch-flex-upload'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ xml }),
  })
  const j = await r.json().catch(() => ({}))
  return {
    ok: Boolean((j as any).ok) && r.ok,
    error: (j as any).error,
    count: (j as any).count,
    updated_accounts: (j as any).updated_accounts,
    message: (j as any).message,
  }
}

/** R-A2: Add one execution manually (historical entry). */
export async function createExecution(body: Record<string, unknown>): Promise<{ ok: boolean; account_executions_id?: number; error?: string }> {
  const r = await fetch(tradingUrl('/executions'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  const j = await r.json().catch(() => ({}))
  const ok = Boolean((j as any).ok) && r.ok
  const detail = (j as any).detail
  const detailMsg =
    typeof detail === 'string'
      ? detail
      : Array.isArray(detail) && detail[0]?.msg
        ? detail[0].msg
        : undefined
  const statusMsg = `${r.status} ${r.statusText || ''}`.trim()
  const error = (j as any).error || detailMsg || (!r.ok ? statusMsg : undefined)
  return { ok, account_executions_id: (j as any).account_executions_id, error }
}

/** R-A2: Update one execution by account_executions_id (manual correction). */
export async function updateExecution(account_executions_id: number, body: Record<string, unknown>): Promise<{ ok: boolean; error?: string }> {
  const r = await fetch(tradingUrl(`/executions/${account_executions_id}`), {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  const j = await r.json().catch(() => ({}))
  const ok = Boolean((j as any).ok) && r.ok
  const detail = (j as any).detail
  const detailMsg =
    typeof detail === 'string'
      ? detail
      : Array.isArray(detail) && detail[0]?.msg
        ? detail[0].msg
        : undefined
  const statusMsg = `${r.status} ${r.statusText || ''}`.trim()
  const error = (j as any).error || detailMsg || (!r.ok ? statusMsg : undefined)
  return { ok, error }
}

/** R-A2: Delete one execution by account_executions_id. */
export async function deleteExecution(account_executions_id: number): Promise<{ ok: boolean; error?: string }> {
  const r = await fetch(tradingUrl(`/executions/${account_executions_id}`), { method: 'DELETE' })
  const j = await r.json().catch(() => ({}))
  const ok = Boolean((j as any).ok) && r.ok
  const detail = (j as any).detail
  const detailMsg =
    typeof detail === 'string'
      ? detail
      : Array.isArray(detail) && detail[0]?.msg
        ? detail[0].msg
        : undefined
  const statusMsg = `${r.status} ${r.statusText || ''}`.trim()
  const error = (j as any).error || detailMsg || (!r.ok ? statusMsg : undefined)
  return { ok, error }
}

/** POST /executions/option-stock-links/query — bulk load links for many option ids (by account batches). */
export async function postOptionStockLinksQuery(body: {
  batches: Array<{ account_id: string; option_account_executions_ids: number[] }>
}): Promise<{ by_option_id: Record<string, OptionStockLinkSummary>; error?: string }> {
  const r = await fetch(tradingUrl('/executions/option-stock-links/query'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  const j = await r.json().catch(() => ({}))
  if (!r.ok) {
    return { by_option_id: {}, error: (j as { error?: string }).error || r.statusText }
  }
  return {
    by_option_id: (j as { by_option_id?: Record<string, OptionStockLinkSummary> }).by_option_id ?? {},
    error: (j as { error?: string }).error,
  }
}

/** GET /executions/option-stock-links — stock legs linked to an OPT row + slippage total. */
export async function fetchOptionStockLinks(
  account_id: string,
  option_account_executions_id: number,
): Promise<{ links: OptionStockLinkRow[]; slippage_total: number | null; error?: string }> {
  const q = new URLSearchParams()
  q.set('account_id', account_id.trim())
  q.set('option_account_executions_id', String(option_account_executions_id))
  const r = await fetch(tradingUrl(`/executions/option-stock-links?${q}`))
  const j = await r.json().catch(() => ({}))
  if (!r.ok) {
    return { links: [], slippage_total: null, error: (j as { error?: string }).error || r.statusText }
  }
  return {
    links: (j as { links?: OptionStockLinkRow[] }).links ?? [],
    slippage_total: (j as { slippage_total?: number | null }).slippage_total ?? null,
    error: (j as { error?: string }).error,
  }
}

/** GET /executions/stock-link-candidates — STK fills for underlying (performance book). */
export async function fetchStockLinkCandidates(params: {
  account_id: string
  option_account_executions_id: number
  trade_date_from?: string
  trade_date_to?: string
  limit?: number
}): Promise<{
  executions: Execution[]
  underlying_symbol?: string
  trade_date_from?: string
  trade_date_to?: string
  error?: string
}> {
  const q = new URLSearchParams()
  q.set('account_id', params.account_id.trim())
  q.set('option_account_executions_id', String(params.option_account_executions_id))
  if (params.trade_date_from?.trim()) q.set('trade_date_from', params.trade_date_from.trim())
  if (params.trade_date_to?.trim()) q.set('trade_date_to', params.trade_date_to.trim())
  if (params.limit != null) q.set('limit', String(params.limit))
  const r = await fetch(tradingUrl(`/executions/stock-link-candidates?${q}`))
  const j = await r.json().catch(() => ({}))
  if (!r.ok) {
    return { executions: [], error: (j as { error?: string }).error || r.statusText }
  }
  return {
    executions: (j as { executions?: Execution[] }).executions ?? [],
    underlying_symbol: (j as { underlying_symbol?: string }).underlying_symbol,
    trade_date_from: (j as { trade_date_from?: string }).trade_date_from,
    trade_date_to: (j as { trade_date_to?: string }).trade_date_to,
    error: (j as { error?: string }).error,
  }
}

/** POST /executions/option-stock-links */
export async function createOptionStockLink(body: {
  account_id: string
  option_account_executions_id: number
  stock_account_executions_id: number
  role?: string | null
  note?: string | null
}): Promise<{ ok: boolean; link_id?: number | null; error?: string; warning?: string | null }> {
  const r = await fetch(tradingUrl('/executions/option-stock-links'), {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  })
  const j = await r.json().catch(() => ({}))
  return {
    ok: Boolean((j as any).ok) && r.ok,
    link_id: (j as any).link_id,
    error: (j as any).error,
    warning: (j as any).warning ?? null,
  }
}

/** DELETE /executions/option-stock-links/{link_id} */
export async function deleteOptionStockLink(
  link_id: number,
  account_id: string,
): Promise<{ ok: boolean; error?: string }> {
  const q = new URLSearchParams()
  q.set('account_id', account_id.trim())
  const r = await fetch(tradingUrl(`/executions/option-stock-links/${link_id}?${q}`), { method: 'DELETE' })
  const j = await r.json().catch(() => ({}))
  return { ok: Boolean((j as any).ok) && r.ok, error: (j as any).error }
}
